provider/core/src/request.rs

```diff
@@ -579,170 +698,94 @@ impl DataLocale {
-        };
-        &DEFAULT
+        DataMarkerAttributes::empty()
     }
 }
 
 impl Deref for DataMarkerAttributes {
     type Target = str;
     #[inline]
     fn deref(&self) -> &Self::Target {
-        match &self.value {
-            DataMarkerAttributesInner::Empty => "",
-            DataMarkerAttributesInner::Boxed(s) => s.deref(),
-            DataMarkerAttributesInner::Stack(s) => s.as_str(),
-        }
+        &self.value
     }
 }
 
-impl PartialEq for DataMarkerAttributes {
+impl Debug for DataMarkerAttributes {
     #[inline]
-    fn eq(&self, other: &Self) -> bool {
-        self.deref() == other.deref()
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.value.fmt(f)
     }
 }
 
-impl Eq for DataMarkerAttributes {}
-
-impl PartialOrd for DataMarkerAttributes {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
-    }
-}
+/// TODO
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct AttributeParseError;
 
-impl Ord for DataMarkerAttributes {
-    fn cmp(&self, other: &Self) -> Ordering {
-        self.deref().cmp(other.deref())
+impl DataMarkerAttributes {
+    const fn validate(s: &str) -> Result<(), AttributeParseError> {
+        let mut i = 0;
+        while i < s.len() {
+            #[allow(clippy::indexing_slicing)] // duh
+            if !matches!(s.as_bytes()[i], b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'-' | b'_') {
+                return Err(AttributeParseError);
+            }
+            i += 1;
+        }
+        Ok(())
     }
-}
 
-impl Debug for DataMarkerAttributes {
-    #[inline]
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.deref().fmt(f)
-    }
-}
+    /// TODO
+    pub const fn try_from_str(s: &str) -> Result<&Self, AttributeParseError> {
+        let Ok(()) = Self::validate(s) else {
+            return Err(AttributeParseError);
+        };
 
-impl Hash for DataMarkerAttributes {
-    #[inline]
-    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
-        self.deref().hash(state)
+        Ok(unsafe { &*(s as *const str as *const DataMarkerAttributes) })
     }
-}
 
-impl FromStr for DataMarkerAttributes {
-    type Err = core::convert::Infallible;
+    /// TODO
+    pub fn try_from_string(s: String) -> Result<Box<Self>, AttributeParseError> {
+        let Ok(()) = Self::validate(&s) else {
+            return Err(AttributeParseError);
+        };
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        if s.is_empty() {
-            Ok(Self {
-                value: DataMarkerAttributesInner::Empty,
-            })
-        } else if let Ok(tiny) = s.parse() {
-            Ok(Self {
-                value: DataMarkerAttributesInner::Stack(tiny),
-            })
-        } else {
-            Ok(Self {
-                value: DataMarkerAttributesInner::Boxed(s.into()),
-            })
-        }
+        Ok(unsafe { core::mem::transmute(s.into_boxed_str()) })
```

Comment:
Issue, here and above: Needs safety comment

Another use case for https://github.com/unicode-org/icu4x/pull/5101 / https://github.com/unicode-org/icu4x/issues/2310 / https://github.com/unicode-org/icu4x/issues/5127